3단계 : 질문 삭제하기 리팩터링

tyranotyrano opened the pull request:
안녕하세요😊 2주차 JPA 미션 진행자 최영진입니다.
3단계 질문 삭제하기 리팩터링 미션 리뷰 요청 드리겠습니다!
3단계 진행 중 궁금한 내용은 아래에 정리하였습니다.

---
Q1. **(답변 완료)** 이번 리팩토링 미션을 진행하면서 한번에 모두 리팩토링 하지 않고 Question 삭제 -> Answer 삭제 -> DeleteHistory 생성 순으로 단계적으로 리팩토링을 진행하였습니다. 리팩토링 하는 과정에서 각 단계별로 비즈니스 로직 및 테스트 코드를 정삭적으로 돌아가게 하기위해 임시적으로 getter 메소드를 추가하였습니다. 그리고 모든 리팩토링 완료 후 불필요한 getter 를 삭제하였습니다. 리팩토링 과정에서 발생하는 임시 getter 메소드들은 모든 처리가 완료된 후 지운다는 전제하에 편의를 위해 추가해도 괜찮을까요? 아니면 리팩토링 과정 단계에서부터도 엄격하게 사용을 최소화 하는게 좋을까요?

Q2. **(답변 완료)** 처음엔 Cascade 나 orphanRemoval 옵션을 주어 Question 이 삭제되면 Answer 도 영속성 전이가 일어나 자동으로 삭제되도록 설계를 했었습니다. 하지만 이렇게 개발을 하니 Answer 의 삭제 조건을 통과하지 않고 그냥 Answer 를 삭제하는 문제가 발생하였습니다. 아직 Cascade 나 orphanRemoval 사용이 익숙치 않아서 이런 옵션들을 사용할 때 주의해야할 점이나 사용 시 팁같은 것들이 있다면 공유받고싶습니다.😃

--- 
3단계 피드백 반영 완료하였습니다. 다시 리뷰 부탁드리겠습니다.😁👍

tyranotyrano committed:
refactor(qna.domain package): 불필요한 Cascade 속성 제거 및 Lazy Loading 속성 적용

불필요한 Cascade 속성 제거 및 Lazy Loading 속성을 적용한다.
- Answer : 불필요한 Cascade 속성 제거 및 Question 에 FetchType.LAZY 적용
- Question : 불필요한 Cascade 속성 제거

tyranotyrano committed:
docs(README.md): 3단계 질문 삭제하기 리팩토링 기능 요구사항 작성

3단계 질문 삭제하기 리팩토링 기능 요구사항을 작성한다.
- README.md : 3단계 질문 삭제하기 리팩토링 기능 요구사항 작성

tyranotyrano committed:
refactor(QnaService.java): Question 삭제 로직 리팩토링

QnaService 의 Question 삭제 로직을 Question 도메인 모델로 이동하는 리팩토링을 한다.
- QnaService : Question 삭제 로직 일부를 Question 도메인 모델로 이동
- Question : Question 삭제 로직 구현
- QnaServiceTest : 답변이 없는 질문이 정상적으로 삭제되는 테스트 코드 추가
- README.md : 완료된 기능 요구사항 체크

tyranotyrano committed:
refactor(QnaService.java): Answer 삭제 로직 리팩토링

QnaService 의 Answer 삭제 로직을 Answer 도메인 모델로 이동하는 리팩토링을 한다.
- QnaService : Answer 삭제 로직을 Answer 도메인 모델로 이동
- Answer : Answer 삭제 로직 구현
- Answers : List<Answer> 를 담고있는 1급 컬렉션
- Question : Answers 와 연관관계 매핑
- QnaServiceTest : Question 의 Answers 연관관계 매핑으로 인해 불필요한 코드 제거
- README.md : 완료된 기능 요구사항 체크

tyranotyrano committed:
refactor(QnaService.java): DeleteHistory 생성 로직 리팩토링

QnaService 의 DeleteHistory 생성 로직을 Question, Answer 도메인 모델로 이동하는 리팩토링을 한다.
- QnaService : DeleteHistory 생성 로직을 Question, Answer 도메인 모델로 이동
- Question : Question 삭제기록용 DeleteHistory 생성 로직 구현
- Answer : Answer 삭제기록용 DeleteHistory 생성 로직 구현
- Answers : 삭제되는 모든 Answer 에 대해 삭제기록용 DeleteHistory 생성 로직 구현
- DeleteHistory : Question, Answer 삭제기록용 팩토리 메소드 추가
- DeleteHistories : List<DeleteHistory> 를 담고있는 1급 컬렉션
- QnaServiceTest : DeleteHistory 생성 팩토리 메소드 적용
- README.md : 완료된 기능 요구사항 체크

tyranotyrano committed:
test(QuestionTest.java): Question 삭제 테스트 코드 추가

Question 삭제 테스트 코드를 추가한다.
- QuestionTest : Question 삭제 테스트 코드 추가
- Question : Answers 의 타입을 final 로 수정

tyranotyrano committed:
test(AnswerTest.java): Answer 삭제 테스트 코드 추가

Answer 삭제 테스트 코드를 추가한다.
- AnswerTest : Question 삭제 테스트 코드 추가
- Answer : validateDeleteAnswerAuthority() 유효성 체크 로직에 "Question 의 Writer 와 Answer 의 Writer 가 다르면 Answer 를 삭제할 수 없다" 는 조건 추가
- Question : isSameWriter() 메소드 추가

tyranotyrano committed:
refactor(DeleteHistoryService.java): DeleteHistoryRepository 를 final 로 수정

DeleteHistoryService 의 DeleteHistoryRepository 를 final 로 수정
- DeleteHistoryService : DeleteHistoryRepository 를 final 로 수정

slamdunk7575 commented on src/main/java/qna/service/QnaService.java:
```diff
@@ -35,24 +33,8 @@ public Question findQuestionById(Long id) {
-            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
-        }
-        deleteHistoryService.saveAll(deleteHistories);
+        DeleteHistories deleteHistories = question.delete(loginUser);
+
+        deleteHistoryService.saveAll(deleteHistories.getValues());
```
깔끔하게 리팩토링 잘해주셨네요 👍

slamdunk7575 commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,41 @@
+
+import qna.CannotDeleteException;
+
+@Embeddable
+public class Answers {
+
+	@OneToMany(mappedBy = "question")
+	private final List<Answer> answers;
```
일급 컬렉션 분리 좋습니다 💯

slamdunk7575 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -83,8 +84,19 @@ public boolean isDeleted() {
+    }
+
+    private void validateDeleteAnswerAuthority(User owner) throws CannotDeleteException {
+        if (isOwner(owner) && question.isSameWriter(writer)) {
```
`question.isSameWriter(writer)`

이부분은 `Question` 에서 검증이 되었으니 제거해도 되지 않을까요?

tyranotyrano replied:
이부분이 제일 고민되는 부분이었는데, Question 을 삭제하는 전반적인 로직에 대해서 Question 에서 이미 유효성 체크를 하였으니 Answer 에서는 패스할 것인가? 아니면 Answer 는 Answer 그 자체로 독립적인 삭제 로직을 검증해야 하는가? 이 두 가지 상황에 대해서 생각을 해봤습니다. 리뷰어님 피드백 처럼 "삭제" 로직 자체가 Question 에서부터 시작하는 로직이고, Answer 를 개별로 삭제하는 요구사항이 없으니 Answer 삭제를 독립적으로 보기보다 Question 삭제 로직과 연결지어서 하나로 보는게 맞을 것 같습니다.😊 refactoring 후 다시 리뷰 요청드리겠습니다! 감사합니다!

slamdunk7575 commented on src/main/java/qna/domain/DeleteHistory.java:
```diff
@@ -34,13 +34,21 @@
         this.createDate = createDate;
     }
 
+    public static DeleteHistory ofQuestion(Long contentId, User deletedByUser) {
+        return new DeleteHistory(ContentType.QUESTION, contentId, deletedByUser, LocalDateTime.now());
+    }
+
+    public static DeleteHistory ofAnswer(Long contentId, User deletedByUser) {
+        return new DeleteHistory(ContentType.ANSWER, contentId, deletedByUser, LocalDateTime.now());
+    }
```
정적 팩토리 메소드 잘 활용해주셨네요 👍

slamdunk7575 commented on src/main/java/qna/domain/Question.java:
```diff
@@ -13,6 +15,8 @@
 
+import qna.CannotDeleteException;
+
 @Entity
 public class Question extends BaseTimeEntity {
 
```
> 모든 원시 값과 문자열을 포장한다.

`예: title` 을 객체로 포장하고 길이에 대한 검증을 해볼 수 있겠네요.
필요하다고 생각되는 부분에 적용해보시면 좋을 것 같아요.

tyranotyrano replied:
JPA 에만 집중해서 개발하다보니 원시값, 문자열 포장을 놓친 것 같습니다😂
이부분도 같이 refactoring 하여 다시 리뷰요청 드리겠습니다!

slamdunk7575 commented on src/test/java/qna/domain/QuestionTest.java:
```diff
@@ -94,4 +103,54 @@ void find2() {
+        assertTrue(Q1.isDeleted());
+    }
+
+    @DisplayName("로그인한 유저가 Question 의 모든 Answer 의 작성자가 아니면 Question 을 삭제할 수 없다.")
+    @Test
+    void delete4() throws CannotDeleteException {
+        // given
+        Question question = questionRepository.save(Q1);
+        Answer answer1 = answerRepository.save(new Answer(user1, question, "Answers Contents1"));
+        Answer answer2 = answerRepository.save(new Answer(user2, question, "Answers Contents2"));
+
+        question.addAnswer(answer1);
+        question.addAnswer(answer2);
+
+        // when
+        assertThatThrownBy(() -> Q1.delete(user2)).isInstanceOf(CannotDeleteException.class);
+    }
```
현재 Repository와 Domain에 대한 테스트가 작성되어 있긴 하지만
리팩토링을 통해 비지니스 로직이 도메인으로 옮겨졌으니 
필요하다고 생각되는 `도메인에(예: Answers) 단위 테스트`를 해보면 좋을 것 같아요.
기존 테스트를 Repository와 Domain 테스트로 나눠보셔도 좋습니다 😄

tyranotyrano replied:
이 부분도 JPA 에만 집중하다보니 각 도메인 객체별 테스트 코드를 작성하지 못한 것 같습니다😂
테스트 코드 추가하여 다시 리뷰 요청드리겠습니다! 세심하게 봐주셔서 감사합니다!

slamdunk7575 commented on src/test/java/qna/service/QnaServiceTest.java:
```diff
@@ -45,53 +46,80 @@ public void setUp() {
         List<DeleteHistory> deleteHistories = Arrays.asList(
-                new DeleteHistory(ContentType.QUESTION, question.getId(), question.getWriter(), LocalDateTime.now()),
-                new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now())
+                DeleteHistory.ofQuestion(question.getId(), question.getWriter()),
+                DeleteHistory.ofAnswer(answer.getId(), answer.getWriter())
```
> Q1. 이번 리팩토링 미션을 진행하면서 한번에 모두 리팩토링 하지 않고 Question 삭제 -> Answer 삭제 -> DeleteHistory 생성 순으로 단계적으로 리팩토링을 진행하였습니다. 리팩토링 하는 과정에서 각 단계별로 비즈니스 로직 및 테스트 코드를 정삭적으로 돌아가게 하기위해 임시적으로 getter 메소드를 추가하였습니다. 그리고 모든 리팩토링 완료 후 불필요한 getter 를 삭제하였습니다. 리팩토링 과정에서 발생하는 임시 getter 메소드들은 모든 처리가 완료된 후 지운다는 전제하에 편의를 위해 추가해도 괜찮을까요? 아니면 리팩토링 과정 단계에서부터도 엄격하게 사용을 최소화 하는게 좋을까요?

미션을 하시면서 느끼셨겠지만 보통 비지니스 로직을 처리할 때 객체의 상태를 외부로 가져오는 `getter()`를 사용하면서 코드가 복잡해지는데요. 이런 `getter()` 최대한 없애고 객체에게 메시지를 보내는 과정에서 발생하는 것이기 때문에 괜찮다고 생각합니다. 또한 연습하는 과정이니 결과를 출력하기 위해 (예: DTO 에서) 최대한 마지막에 `getter()`를 사용한다고 생각하고 해보시는 것도 좋겠네요 :)

tyranotyrano replied:
답변 감사합니다! getter/setter 를 극단적으로 안쓰려고 하다보니 오히려 리팩토링 하는 과정에서 문제가 많았고, 뭔가 getter 를 사용하면 심적으로 미션을 잘못된 방향으로 가져가는게 아닌가 걱정을 많이 했었습니다. 이번 미션진행하면서 그리고 리뷰어님의 피드백을  곰곰히 생각해보니 getter/setter 를 지양하는 것은 맞지만 꼭 필요한 경우에 제한적으로 getter 를 쓰고 최대한 도메인 객체에게 메시지를 보내어 처리할 수 있게 만드는게 좋다고 판단됩니다!😁

slamdunk7575 commented on src/main/java/qna/domain/Answers.java:
```diff
@@ -0,0 +1,41 @@
+
+import qna.CannotDeleteException;
+
+@Embeddable
+public class Answers {
+
+	@OneToMany(mappedBy = "question")
```
> Q2. 처음엔 Cascade 나 orphanRemoval 옵션을 주어 Question 이 삭제되면 Answer 도 영속성 전이가 일어나 자동으로 삭제되도록 설계를 했었습니다. 하지만 이렇게 개발을 하니 Answer 의 삭제 조건을 통과하지 않고 그냥 Answer 를 삭제하는 문제가 발생하였습니다. 아직 Cascade 나 orphanRemoval 사용이 익숙치 않아서 이런 옵션들을 사용할 때 주의해야할 점이나 사용 시 팁같은 것들이 있다면 공유받고싶습니다.😃

객체의 `생성과 소멸`을 같이 하는 경우에 `Cascade` 설정을 고려해볼 수 있을 것 같아요.
알고 계시겠지만 `Cascade` 옵션에는 다양한 값들이 있기 때문에 상황에 맞춰 설정을 해보시면 좋을 것 같네요.

tyranotyrano replied:
Cascade 를 사용할 때 습관적으로 항상 Cascade.ALL 을 사용했었는데, 피드백 주신 것 처럼 하위 엔티티가 상위 엔티티 1군데 에서만 종속적이고 비즈니스 로직으로 상위 엔티티가 삭제되면 모든 하위 엔티티가 삭제된다는 요구사항이 있다면 Cascade 를 좀더 활용할 수 있을 것 같습니다. 2단계에서 공유해주신 내용 덕분에 Cascade, orphanRemoval 옵션에 대해 많은 공부가 되었습니다. 감사합니다!👍

slamdunk7575 reviewed (changes requested):
질문 삭제하기 리팩토링 정말 잘해주셨네요.
질문에 대한 답변과 소소한 코멘트 몇가지 남겨드렸습니다.
미션 마무리까지 얼마남지 않으셨네요~ 💪

tyranotyrano committed:
refactor(Answer.java): Answer 삭제 관련 불필요한 유효성 검증 로직 제거

Answer 삭제 관련 불필요한 유효성 검증 로직을 제거한다.
- Answer : Answer 삭제 관련 불필요한 유효성 검증 로직 제거
- Question : 불필요한 isSameWriter() 메소드 제거
- AnswerTest : 불필요한 테스트 코드 제거

tyranotyrano committed:
refactor(AnswerContents.java): Answer 의 contents 를 포장하는 AnswerContents 도메인 모델 추가

Answer 의 contents 를 포장하는 AnswerContents 도메인 모델을 추가한다.
- AnswerContents : contents 를 포장하는 도메인 모델
- Answer : contents -> AnswerContents 로 변경
- AnswerContentsTest : AnswerContents 테스트 코드 추가

tyranotyrano committed:
refactor(Contents.java): AnswerContents -> Contents 로 클래스 이름 변경

AnswerContents -> Contents 로 클래스 이름을 변경한다.
- Contents : AnswerContents -> Contents 로 클래스 이름 변경
- Answer, AnswerContentsTest : AnswerContents -> Contents 로 변경

tyranotyrano committed:
refactor(QuestionTitle.java): Question 의 title 를 포장하는 QuestionTitle 도메인 모델 추가

Question 의 title 를 포장하는 QuestionTitle 도메인 모델을 추가한다.
- QuestionTitle : Question 의 title 을 포장하는 도메인 모델
- Question : title -> QuestionTitle, contents -> Contents 도메인 모델 적용
- Contents : isEmpty() 메소드 위치 수정
- QuestionTitleTest : QuestionTitle 테스트 코드 추가

tyranotyrano committed:
test(AnswersTest.java): Answers 테스트 코드 추가

Answers 테스트 코드를 추가한다.
- AnswersTest : Answers 테스트 코드 추가
- Answers : isAllDeleted() 메소드 추가

slamdunk7575 commented on src/main/java/qna/domain/QuestionTitle.java:
```diff
@@ -0,0 +1,52 @@
+import javax.persistence.Column;
+import javax.persistence.Embeddable;
+
+@Embeddable
+public class QuestionTitle {
```
원시값 포장 💯

slamdunk7575 commented on src/main/java/qna/domain/Answer.java:
```diff
@@ -83,8 +84,19 @@ public boolean isDeleted() {
+        return DeleteHistory.ofAnswer(id, owner);
+    }
+
+    private void validateDeleteAnswerAuthority(User owner) throws CannotDeleteException {
+        if (isOwner(owner)) {
+            return;
+        }
+
+        throw new CannotDeleteException("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.");
     }
```
피드백 반영 잘해주셨네요 👍

slamdunk7575 commented on src/test/java/qna/domain/AnswersTest.java:
```diff
@@ -0,0 +1,92 @@
+		question = questionRepository.save(QuestionTest.Q1.writeBy(user1));
+	}
+
+	@DisplayName("Answers 에서 전체 삭제(deleteAll())를 하면 모든 Answer 는 삭제상태(deleted=true) 로 설정된다.")
+	@Test
+	void deleteAll1() throws CannotDeleteException {
```
현재는 도메인에 비지니스 로직이 많지 않지만
남은 미션들에서는 순수 도메인에 대한 테스트를 별도로 만들어 보면 좋을 것 같아요 🙂

slamdunk7575 reviewed (approved):
피드백 열심히 반영해주셨네요.
제가 드렸던 피드백들이 앞으로 영진님께 조금이라도 도움이 되었으면 좋겠네요.
그동안 고생 많으셨습니다. 감사합니다. 🙇
남은 미션도 꾸준히 진행하시고 끝까지 완주하시길 응원하겠습니다! 🎉
